powerauth-test-server/src/main/java/com/wultra/security/powerauth/app/testserver/service/OperationsService.java

```diff
@@ -210,15 +212,69 @@ public Response approveOperation(OperationApproveInternalRequest request) throws
+     * @return Operation approval response.
+     * @throws RemoteExecutionException In case internal calls fail.
+     * @throws SignatureVerificationException In case signature verification fails.
+     * @throws ActivationFailedException In case activation is not found.
+     * @throws AppConfigNotFoundException In case app configuration is not found.
+     */
+    public Response rejectOperation(OperationRejectInternalRequest request) throws AppConfigNotFoundException, ActivationFailedException, SignatureVerificationException, RemoteExecutionException {
+        final String applicationId = request.getApplicationId();
+        final TestConfigEntity appConfig = getTestAppConfig(applicationId);
+        final JSONObject resultStatusObject = resultStatusUtil.getTestStatus(request.getActivationId());
+
+        final String operationId = request.getOperationId();
+        final String reason = request.getReason();
+
+        final Map<String, String> map = new HashMap<>();
+        map.put("id", operationId);
+        if (reason != null) {
+            map.put("reason", reason);
+        }
+
+        final byte[] payload;
+        try {
+            String payloadString = new ObjectMapper().writeValueAsString(new ObjectRequest<>(map));
+            payload = payloadString.getBytes(StandardCharsets.UTF_8);
+        } catch (JsonProcessingException e) {
+            throw new SignatureVerificationException("Unable to serialize data");
+        }
 
+        final VerifySignatureStepModel model = new VerifySignatureStepModel();
+        model.setData(payload);
+        model.setResourceId("/operation/cancel");
+        model.setUriString(config.getEnrollmentServiceUrl() + "/api/auth/token/app/operation/cancel");
+        model.setHttpMethod("POST");
+        model.setApplicationKey(appConfig.getApplicationKey());
+        model.setApplicationSecret(appConfig.getApplicationSecret());
+        model.setSignatureType(PowerAuthSignatureTypes.POSSESSION);
+        model.setVersion(config.getVersion());
+        model.setResultStatusObject(resultStatusObject);
+
+        boolean success = false;
+        try {
+            final ObjectStepLogger stepLogger = new ObjectStepLogger();
+            verifySignatureStep.execute(stepLogger, model.toMap());
+            for (StepItem item: stepLogger.getItems()) {
+                StepItemLogger.log(logger, item);
+                if ("Signature verified".equals(item.name())) {
+                    success = true;
+                }
+            }
+        } catch (Exception ex) {
+            logger.warn("Remote execution failed, reason: {}", ex.getMessage());
+            logger.debug(ex.getMessage(), ex);
+            throw new RemoteExecutionException("Remote execution failed");
+        }
+
+        resultStatusUtil.persistResultStatus(resultStatusObject);
+
+        if (!success) {
+            throw new SignatureVerificationException("Signature verification failed");
+        }
```

Comment:
You may use Java 8+ API.

```suggestion
        final ObjectStepLogger stepLogger;
        try {
            stepLogger = new ObjectStepLogger();
            verifySignatureStep.execute(stepLogger, model.toMap());
            stepLogger.getItems()
                    .forEach(it -> StepItemLogger.log(logger, it));
        } catch (Exception ex) {
            logger.warn("Remote execution failed, reason: {}", ex.getMessage());
            logger.debug(ex.getMessage(), ex);
            throw new RemoteExecutionException("Remote execution failed");
        }

        resultStatusUtil.persistResultStatus(resultStatusObject);

        stepLogger.getItems().stream()
                .filter(item -> "Signature verified".equals(item.name()))
                .findAny()
                .orElseThrow(() -> new SignatureVerificationException("Signature verification failed"));
```